pkg/clusteragent/admission/mutate/common.go

```diff
@@ -153,3 +159,67 @@ func containsVolumeMount(volumeMounts []corev1.VolumeMount, element corev1.Volum
 	}
 	return false
 }
+
+// shouldInject returns true if Admission Controller should inject standard tags, APM configs and APM libraries
+func shouldInject(pod *corev1.Pod) bool {
+	// If a pod explicitly sets the label admission.datadoghq.com/enabled, make a decision based on its value
+	if val, found := pod.GetLabels()[admCommon.EnabledLabelKey]; found {
+		switch val {
+		case "true":
+			return true
+		case "false":
+			return false
+		default:
+			log.Warnf("Invalid label value '%s=%s' on pod %s should be either 'true' or 'false', ignoring it", admCommon.EnabledLabelKey, val, podString(pod))
+		}
+	}
+
+	return isApmInstrumentationEnabled(pod.GetNamespace()) || config.Datadog.GetBool("admission_controller.mutate_unlabelled")
+}
+
+// isApmInstrumentationEnabled indicates if Single Step Insstrumentation is enabled for the namespace in the cluster
+// Single Step Instrumentation is enabled if:
+//   - cluster-wide configuration `apm_config.instrumentation.enabled` is true and the namespace is not excluded via `apm_config.instrumentation.disabled_namespaces`
+//   - cluster-wide configuration `apm_config.instrumentation.enabled` is false and the namespace is whitelisted via `apm_config.instrumentation.disabled_namespaces`
+func isApmInstrumentationEnabled(namespace string) bool {
+	apmInstrumentationEnabled := config.Datadog.GetBool("apm_config.instrumentation.enabled")
+	isNsEnabled := apmInstrumentationEnabled
+
+	apmEnabledNamespaces := config.Datadog.GetStringSlice("apm_config.instrumentation.enabled_namespaces")
+	apmDisabledNamespaces := config.Datadog.GetStringSlice("apm_config.instrumentation.disabled_namespaces")
+
+	// If Single Step Instrumentation is enabled in the cluster, enabling it in the specific namespaces is redundant
+	if apmInstrumentationEnabled && len(apmEnabledNamespaces) > 0 {
+		log.Warnf("Redundant configuration option `apm_config.instrumentation.enabled_namespaces:%v`", apmEnabledNamespaces)
+	}
+
+	// If Single Step Instrumentation is disabled in the cluster, disabling it in the specific namespaces is redundant
+	if !apmInstrumentationEnabled && len(apmDisabledNamespaces) > 0 {
+		log.Warnf("Redundant configuration option `apm_config.instrumentation.disabled_namespaces:%v`", apmDisabledNamespaces)
+	}
+
+	// apm.instrumentation.enabled_namespaces and apm.instrumentation.disabled_namespaces configuration cannot be set at the same time
```

Comment:
One way to simplify that's usually used in the Agent to `include/exclude` is to support glob and have `enabled > disabled`
```
disabled: *
enabled: myappns
```

It would also allow you to easily add `kube-system` in `disabled`.

Reply:
Do you have any examples of this logic in the agent?

Reply:
Yes currently most `include/exclude` are using this syntax: https://docs.datadoghq.com/containers/guide/autodiscovery-management/?tab=datadogoperator

Which is implemented by this code: https://github.com/DataDog/datadog-agent/blob/0a5b37375255bc0f6e8aa19c180d3095198e603a/pkg/util/containers/filter.go